src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,297 @@
+class PromiseEndpoint {
+ public:
+  PromiseEndpoint(
+      std::unique_ptr<grpc_event_engine::experimental::EventEngine::Endpoint>
+          endpoint,
+      SliceBuffer already_received);
+  ~PromiseEndpoint();
+
+  // Returns a promise that resolves to a `absl::Status` indicating the result
+  // of the write operation.
+  //
+  // Concurrent writes are not supported, which means callers should not call
+  // `Write()` before the previous write finishes. Doing that results in
+  // undefined behavior.
+  auto Write(SliceBuffer data) {
+    {
+      MutexLock lock(&write_mutex_);
+      // Assert previous write finishes.
+      GPR_ASSERT(!write_result_.has_value());
+      // TODO(ladynana): Replace this with `SliceBufferCast<>` when it is
+      // available.
+      grpc_slice_buffer_swap(write_buffer_.c_slice_buffer(),
+                             data.c_slice_buffer());
+    }
+    // If `Write()` returns true immediately, the callback will not be called.
+    // We still need to call our callback to pick up the result.
+    if (endpoint_->Write(std::bind(&PromiseEndpoint::WriteCallback, this,
+                                   std::placeholders::_1),
+                         &write_buffer_,
+                         nullptr /* uses default arguments */)) {
+      WriteCallback(absl::OkStatus());
+    }
+    return [this]() -> Poll<absl::Status> {
+      MutexLock lock(&write_mutex_);
+      // If current write isn't finished return `Pending()`, else return write
+      // result.
+      if (!write_result_.has_value()) {
+        write_waker_ = Activity::current()->MakeNonOwningWaker();
+        return Pending();
+      } else {
+        const auto ret = *write_result_;
+        write_result_.reset();
+        return ret;
+      }
+    };
+  }
+
+  // Returns a promise that resolves to `SliceBuffer` with
+  // `num_bytes` bytes.
+  //
+  // Concurrent reads are not supported, which means callers should not call
+  // `Read()` before the previous read finishes. Doing that results in
+  // undefined behavior.
+  auto Read(size_t num_bytes) {
+    ReleasableMutexLock lock(&read_mutex_);
+    // Assert previous read finishes.
+    GPR_ASSERT(!read_result_.has_value());
+    // Should not have pending reads.
+    GPR_ASSERT(pending_read_buffer_.Count() == 0u);
+    if (read_buffer_.Length() < num_bytes) {
+      lock.Release();
+      // Set read args with hinted bytes.
+      grpc_event_engine::experimental::EventEngine::Endpoint::ReadArgs
+          read_args;
+      read_args.read_hint_bytes = num_bytes;
+      // If `Read()` returns true immediately, the callback will not be
+      // called. We still need to call our callback to pick up the result and
+      // maybe do further reads.
+      if (endpoint_->Read(std::bind(&PromiseEndpoint::ReadCallback, this,
+                                    std::placeholders::_1, num_bytes,
+                                    absl::nullopt /* uses default arguments */),
```

Comment:
The current implementation makes `requested_read_args` in `PromiseEndpoint::ReadCallback()` useless. We probably should consider clean it up.

The role of `requested_read_args` in the prototype PR is based on the assumption that the `num_byte` in this function could be large so that a single call to `endpoint_->Read()` will not be enough in most cases. Smaller `num_byte` should call `ReadSlice()` instead. The argument `requested_read_args` is the only difference for `Read()` and `ReadSlice()` when creating the lambda in the prototype PR.

Seeing the implementation here, I am a little confused of the difference between `Read()` and `ReadSlice()`.

Reply:
@Vignesh2208 correct me if I am wrong, from my understanding, read_hint_bytes is a good to have setting, but not guarantee that EventEngine will read all the bytes at once? (Also, I should use `read_args` instead of `absl::nullopt` in the bind function as well, I will send a following PR to fix this.)

I think the difference between Read() and ReadSlice() is mostly the return type and use scenarios. And since ReadSlice() will do a copy from SliceBuffer to Slice, we will probably avoid using it for now until we find a better solution to remove the memcpy issue.

Reply:
If we always have `read_hint_bytes`, we don't need to have the argument in the callback. You may clean that up in the following-up PR as well.

Reply:
Oh, I see what you meant now. Yeah, the parameter of `requested_read_args` in ReadCallback isn't useful anymore, I will clean it up.